File: Fibonacci Heaps/fibonacci_delete_node.py
# File: fibonacci_delete_node.py
# Program: Delete a node in a Fibonacci Heap

class Node:
    def __init__(self, key):
        self.key = key
        self.left = self
        self.right = self


class FibonacciHeap:

    def __init__(self):
        self.min = None

    def insert(self, key):

        node = Node(key)

        if self.min is None:
            self.min = node
        else:

            node.left = self.min
            node.right = self.min.right
            self.min.right.left = node
            self.min.right = node

            if node.key < self.min.key:
                self.min = node

        return node

    def decrease_key(self, node, new_key):

        node.key = new_key

        if node.key < self.min.key:
            self.min = node

    def extract_min(self):

        if self.min is None:
            return

        temp = self.min

        if temp.right == temp:
            self.min = None
        else:
            temp.left.right = temp.right
            temp.right.left = temp.left
            self.min = temp.right

        print("Deleted node with key", temp.key)

    def delete_node(self, node):

        self.decrease_key(node, float('-inf'))
        self.extract_min()


heap = FibonacciHeap()

n1 = heap.insert(10)
n2 = heap.insert(20)

heap.delete_node(n2)

File: Fibonacci Heaps/FibonacciDeleteNode.java
// File: FibonacciDeleteNode.java
// Program: Delete a node in a Fibonacci Heap

class Node {
    int key;
    Node left, right;

    Node(int key) {
        this.key = key;
        left = right = this;
    }
}

public class FibonacciDeleteNode {

    static Node min = null;

    static Node insert(Node min, int key)
    {
        Node n = new Node(key);

        if (min == null)
            return n;

        n.left = min;
        n.right = min.right;
        min.right.left = n;
        min.right = n;

        if (n.key < min.key)
            min = n;

        return min;
    }

    static void decreaseKey(Node x, int newKey)
    {
        x.key = newKey;

        if (x.key < min.key)
            min = x;
    }

    static void extractMin()
    {
        if (min == null)
            return;

        Node temp = min;

        if (temp.right == temp)
            min = null;
        else
        {
            temp.left.right = temp.right;
            temp.right.left = temp.left;
            min = temp.right;
        }

        System.out.println("Deleted node with key " + temp.key);
    }

    static void deleteNode(Node x)
    {
        decreaseKey(x, Integer.MIN_VALUE);
        extractMin();
    }

    public static void main(String[] args)
    {
        Node n1 = new Node(10);
        Node n2 = new Node(20);

        min = insert(min, 10);
        min = insert(min, 20);

        deleteNode(n2);
    }
}
